sdk/communication/azure-communication-phonenumbers/src/main/java/com/azure/communication/phonenumbers/SipRoutingClientBuilder.java

```diff
@@ -0,0 +1,451 @@
+     * @throws NullPointerException If {@code tokenCredential} is {@code null}.
+     */
+    @Override
+    public SipRoutingClientBuilder credential(TokenCredential tokenCredential) {
+        this.tokenCredential = Objects.requireNonNull(tokenCredential, "'tokenCredential' cannot be null.");
+        return this;
+    }
+
+    /**
+     * Set the endpoint and AzureKeyCredential for authorization.
+     *
+     * @param connectionString connection string in the format "endpoint={endpoint_value};accesskey={accesskey_value}"
+     * @return The updated {@link SipRoutingClientBuilder} object.
+     * @throws NullPointerException If {@code connectionString} is {@code null}.
+     */
+    @Override
+    public SipRoutingClientBuilder connectionString(String connectionString) {
+        Objects.requireNonNull(connectionString, "'connectionString' cannot be null.");
+        CommunicationConnectionString connectionStringObject = new CommunicationConnectionString(connectionString);
+        String endpoint = connectionStringObject.getEndpoint();
+        String accessKey = connectionStringObject.getAccessKey();
+        this
+            .endpoint(endpoint)
+            .credential(new AzureKeyCredential(accessKey));
+        return this;
+    }
+
+    /**
+     * Sets the configuration object used to retrieve environment configuration values during building of the client.
+     *
+     * @param configuration Configuration store used to retrieve environment configurations.
+     * @return The updated {@link SipRoutingClientBuilder} object.
+     */
+    @Override
+    public SipRoutingClientBuilder configuration(Configuration configuration) {
+        this.configuration = configuration;
+        return this;
+    }
+
+    /**
+     * Adds a {@link HttpPipelinePolicy pipeline policy} to apply on each request sent.
+     *
+     * <p><strong>Note:</strong> It is important to understand the precedence order of the HttpTrait APIs. In
+     * particular, if a {@link HttpPipeline} is specified, this takes precedence over all other APIs in the trait, and
+     * they will be ignored. If no {@link HttpPipeline} is specified, a HTTP pipeline will be constructed internally
+     * based on the settings provided to this trait. Additionally, there may be other APIs in types that implement this
+     * trait that are also ignored if an {@link HttpPipeline} is specified, so please be sure to refer to the
+     * documentation of types that implement this trait to understand the full set of implications.</p>
+     *
+     * @param policy A {@link HttpPipelinePolicy pipeline policy}.
+     * @return The updated {@link SipRoutingClientBuilder} object.
+     * @throws NullPointerException If {@code policy} is {@code null}.
+     */
+    @Override
+    public SipRoutingClientBuilder addPolicy(HttpPipelinePolicy policy) {
+        this.additionalPolicies.add(Objects.requireNonNull(policy, "'policy' cannot be null."));
```

Comment:
`HttpPipelinePolicy`s have a concepts of both before retry and after retry that will need to be handled here and during building of the client.

[Example of handling during `addPolicy`](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/SearchClientBuilder.java#L336)

[Example of per-call policy adding](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/util/Utility.java#L131)

[Example of per-request and retry policy adding](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/util/Utility.java#L150)

This allows for finer grain handling when constructing a client as you may have a custom policy the keeps track of all network request made which needs to be added per-request/retry and a custom policy of number of client APIs called that should only happen once per client method call (or a few times depending on if the client API makes multiple calls). This made up example would be helpful in troubleshooting networking issues if you're seeing 4 network requests per client API call.

Reply:
https://github.com/Azure/azure-sdk-for-java/pull/26819/commits/2b5858f08801314162458b7e44339a06a353727a Handling separately policies perRequest and perRetry. Please verify.